### ISSUE TITLE
Reduce code redundancy

### ISSUE THREAD

**Issue by daemon** [castorini/Castor#175]
A few issues we should address:
1. Redundant [argument parsers](https://github.com/castorini/Castor/blob/master/vdpwi/__main__.py#L25) throughout the codebase. I think moving to a [hierarchical JSON config system](https://github.com/daemon/argconf/tree/master/examples) makes sense, where we have a single global configuration for all the universal settings, like the learning rate, weight decay, and optimizer, and then inheriting, model-specific configurations. Does that seem like a sane choice?

2. Redundant [LSTM baseline](https://github.com/castorini/Castor/blob/master/lstm_baseline) and [regularization models](https://github.com/castorini/Castor/blob/master/lstm_regularization). Is there a reason we need two models (both named `LSTMBaseline`)? Isn't LSTM-reg just LSTM-baseline with regularization? 

3. Other obviously redundant code [[1]](https://github.com/castorini/Castor/blob/master/common/dataset.py#L39) [[2]](https://github.com/castorini/Castor/tree/master/datasets)

**Comment by achyudhk** [comment]
For 2, no we wouldn't need two models. I did something similar for regularizing KimCNN and it's fine if we just have one model with optional regularization parameters.
